File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/HotRodTypeConverter.scala
package org.infinispan.server.hotrod

import org.infinispan.compat.TypeConverter
import org.infinispan.commons.marshall.jboss.GenericJBossMarshaller
import org.infinispan.context.Flag
import org.infinispan.commons.marshall.Marshaller

/**
 * Hot Rod type converter for compatibility mode.
 *
 * @author Galder Zamarreño
 * @since 5.3
 */
class HotRodTypeConverter extends TypeConverter[AnyRef, AnyRef, AnyRef, AnyRef] {

   // Default marshaller is the one used by the Hot Rod client,
   // but can be configured for compatibility use cases
   private var marshaller: Marshaller = new GenericJBossMarshaller

   override def boxKey(key: AnyRef): AnyRef = unmarshall(key)

   override def boxValue(value: AnyRef): AnyRef = unmarshall(value)

   override def unboxKey(target: AnyRef): Array[Byte] = marshall(target)

   override def unboxValue(target: AnyRef): Array[Byte] = marshall(target)

   override def supportsInvocation(flag: Flag): Boolean =
      if (flag == Flag.OPERATION_HOTROD) true else false

   override def setMarshaller(marshaller: Marshaller) {
      this.marshaller = marshaller
   }

   private def unmarshall(source: AnyRef): AnyRef = {
      source match {
         case bytes: Array[Byte] =>
            marshaller.objectFromByteBuffer(bytes)
         case _ => source
      }
   }

   private def marshall(source: AnyRef): Array[Byte] =
      if (source != null) marshaller.objectToByteBuffer(source) else null

}

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodTypeConverter.java
package org.infinispan.server.hotrod;

import java.io.IOException;

import org.infinispan.commons.CacheException;
import org.infinispan.commons.marshall.Marshaller;
import org.infinispan.commons.marshall.jboss.GenericJBossMarshaller;
import org.infinispan.compat.TypeConverter;
import org.infinispan.context.Flag;

/**
 * Hot Rod type converter for compatibility mode.
 *
 * @author Galder Zamarreño
 * @since 5.3
 */
public class HotRodTypeConverter implements TypeConverter<Object, Object, Object, Object> {

   // Default marshaller is the one used by the Hot Rod client,
   // but can be configured for compatibility use cases
   private Marshaller marshaller;

   public HotRodTypeConverter() {
      this(null);
   }

   public HotRodTypeConverter(Marshaller marshaller) {
      if (marshaller == null) {
         this.marshaller = new GenericJBossMarshaller();
      } else {
         this.marshaller = marshaller;
      }
   }

   @Override
   public Object boxKey(Object key) {
      return unmarshall(key);
   }

   @Override
   public Object boxValue(Object value) {
      return unmarshall(value);
   }

   @Override
   public Object unboxKey(Object target) {
      return marshall(target);
   }

   @Override
   public Object unboxValue(Object target) {
      return marshall(target);
   }

   @Override
   public boolean supportsInvocation(Flag flag) {
      return flag == Flag.OPERATION_HOTROD;
   }

   @Override
   public void setMarshaller(Marshaller marshaller) {
      this.marshaller = marshaller;
   }

   private Object unmarshall(Object source) {
      if (source instanceof byte[]) {
         try {
            return marshaller.objectFromByteBuffer((byte[]) source);
         } catch (IOException | ClassNotFoundException e) {
            throw new CacheException(e);
         }
      } else {
         return source;
      }
   }

   private Object marshall(Object source) {
      if (source != null) {
         try {
            if (marshaller.isMarshallable(source)) {
               return marshaller.objectToByteBuffer(source);
            } else {
               return source;
            }
         } catch (Exception e) {
            throw new CacheException(e);
         }
      } else return null;
   }
}
